### PR TITLE
Add `ReadinessProbeFailureThreshold` field and lower readiness probe  failThreshold.

### LINKED ISSUE #3671: Isn't Pormetheus' readiness probe too slow to detect a failure?
Hello,

The comment here https://github.com/prometheus-operator/prometheus-operator/blob/7daed4b44b1a198972f3d7ea3c904895fb536861/pkg/prometheus/statefulset.go#L652 indicates that 10 minutes are necessary to let Prometheus do data recovery on startup. The problem is that a failure takes 10 minutes to be detected. Unless I'm mistaken, it's not a problem for a pod to fail it's readiness probe during startup, readiness probe only “decides” if a pod is ready to serve traffic or not. If the readiness probe fails, the pod won't be killed, it just won't be part of its service's endpoints. Could we consider lowering the failure threshold for quicker failure detection?

### PR DESCRIPTION
**Pull request by mikechengwei**
Fix #3671 .
Add readiness probe threshold field and lower failThreshold.
**Release Note Template (will be copied)**

```release-note:REPLACEME
Add `ReadinessProbeFailureThreshold` field and lower readiness probe  failThreshold.
```

### CODE REVIEW

**Comment by coderanger** [comment]
Is there a reason to make this an explicit flag instead of using the `containers` deep merge?

**Comment by lilic** [comment]
I don't see any problems with adding this, containers merge is slightly hacky and very hard to discover.
cc @prometheus-operator/prometheus-operator-reviewers any objections to this?

@mikechengwei thanks for your PR, seems like generate is failing do you mind having a look at this. Also do you mind making your commits a bit neater, to describe the files you are touching and the actual field you are adding, as a note we merge as is, we do not squash, something to keep in mind :)

File: pkg/apis/monitoring/v1/types.go [hunk 1]
```diff
@@ -366,6 +366,8 @@ type PrometheusSpec struct {
 	// the desired limit.
 	// Note that if TargetLimit is higher that value will be taken instead.
 	EnforcedTargetLimit *uint64 `json:"enforcedTargetLimit,omitempty"`
+	// Number of readiness probe failure threshold,50 times by default.
```
Comment (lilic) [inline pkg/apis/monitoring/v1/types.go:369]:
```suggestion
	// ReadinessProbeFailureThreshold  is the readiness probe failure threshold.
	// 50 is the default value.
```

File: pkg/prometheus/statefulset.go [hunk 2]
```diff
@@ -606,7 +606,10 @@ func makeStatefulSetSpec(p monitoringv1.Prometheus, c *operator.Config, shard in
 		Handler:          readinessProbeHandler,
 		TimeoutSeconds:   probeTimeoutSeconds,
 		PeriodSeconds:    5,
-		FailureThreshold: 120, // Allow up to 10m on startup for data recovery
+		FailureThreshold: 50, // Allow up to 10m on startup for data recovery
```
Comment (lilic) [inline pkg/prometheus/statefulset.go:609]: What is the reason for this change?

File: pkg/prometheus/statefulset.go [hunk 3]
```diff
@@ -606,7 +606,10 @@ func makeStatefulSetSpec(p monitoringv1.Prometheus, c *operator.Config, shard in
...
+	}
+	if p.Spec.ReadinessProbeFailureThreshold != nil {
```
Comment (lilic) [inline pkg/prometheus/statefulset.go:611]: As @coderanger suggested, this can already be done via container merges, so you have to take that into account. Best is to have a look below this code how we handle container merging for other use cases. Let me know if you have any questions!

**Comment by s-urbaniak** [comment]
No general objections to merge this in, but just two thoughts:

a) we should have a bit better guidance on where we want explicit configuration to be exposed vs. container merging: Potentially we want to expose most critical settings so they are explicitly part of the API. Evaluating a setting as "critical" can be quite subjective at times though.

b) We should generally document container merging a bit better. We describe the term in `api.md` in a few occurences only but lack description of more detailed semantics of the merge strategy.

**Review by paulfantom — CHANGES_REQUESTED** [review]
As far as I can see this is related to https://github.com/prometheus-operator/prometheus-operator/issues/3257, https://github.com/prometheus-operator/prometheus-operator/issues/3587, and https://github.com/prometheus-operator/prometheus-operator/issues/3391. In neither of those we reached any consensus on how to configure probes and all the time our recommendation was to patch containers (I agree with @sur that we should document it more).

I would prefer to have this split into 2 PRs. The one for fixing #3671 (changing hardcoded threshold) and one for exposing a new field in CRD.

**Comment by mikechengwei** [comment]
Ok. I will do it by adding a `patchContainer` field and merge patch field and origin container fields.
Yaml like: 
```
apiVersion: monitoring.coreos.com/v1
kind: Prometheus
metadata:
  labels:
    prometheus: self
  name: self
  namespace: default
spec:
  patchContainer:
    readinessProbe:
      failureThreshold: 20

```
What do you think ? @paulfantom @s-urbaniak

**Comment by mikechengwei** [comment]
@paulfantom Do you mean adding a new field instead of "container", right? E.g.'patchConatiner'.

**Comment by mikechengwei** [comment]
This feature can already be done by merging the conatinenr field and close.
